File: clojure/binary-search/src/binary_search.clj
(ns binary-search)

(defn search-for
  "Returns the index of num in coll, or -1 if num is not found."
  [num coll]
  (loop [low-idx 0 high-idx (dec (count coll))]
    (if (> low-idx high-idx) 
      -1
      (let [mid-idx (quot (+ low-idx high-idx) 2)
            mid-val (nth coll mid-idx)]
        (cond
          (< num mid-val) (recur low-idx (dec high-idx))
          (> num mid-val) (recur (inc low-idx) high-idx)
          :else mid-idx)))))

File: elixir/binary-search/lib/binary_search.ex
defmodule BinarySearch do
  @doc """
    Searches for a key in the tuple using the binary search algorithm.
    It returns :not_found if the key is not in the tuple.
    Otherwise returns {:ok, index}.

    ## Examples

      iex> BinarySearch.search({}, 2)
      :not_found

      iex> BinarySearch.search({1, 3, 5}, 2)
      :not_found

      iex> BinarySearch.search({1, 3, 5}, 5)
      {:ok, 2}

  """

  @spec search(tuple, integer) :: {:ok, integer} | :not_found
  def search({}, _key), do: :not_found
  def search(numbers, key), do: 
    numbers 
    |> Tuple.to_list() 
    |> Enum.sort() 
    |> List.to_tuple() 
    |> do_search(key, 0, tuple_size(numbers) - 1)

  defp do_search(_numbers, _key, min, max) when min > max, do: :not_found
  defp do_search(numbers, key, min, max) do
    mid = div(min + max, 2)
    value = elem(numbers, mid)
    cond do
      value == key -> {:ok, mid}
      value < key -> do_search(numbers, key, mid + 1, max)
      value > key -> do_search(numbers, key, min, mid - 1)
    end
  end
end
